R/dm.R

```diff
@@ -732,56 +732,149 @@ as.list.zoomed_dm <- function(x, ...) {
+  print_glimpse_table_pk(x, table_name, width)
+  print_glimpse_table_fk(x, table_name, width)
+  glimpse(table, width = width, ...)
+}
+
+#' Print table name for a given table in the `dm` object (zoomed or not)
+#' @keywords internal
+#' @noRd
+print_glimpse_table_name <- function(table_name, width) {
+  cat_line("\n", trim_width(paste0("Table: ", tick(table_name)), width))
+}
+
+#' Print details about primary key for a given table in the `dm` object (zoomed or not)
+#' @keywords internal
+#' @noRd
+print_glimpse_table_pk <- function(x, table_name, width) {
+  pk <- dm_get_pk_impl(x, table_name)
+
+  # anticipate that some key columns could have been removed by the user
+  if (is_zoomed(x)) {
+    pk <-
+      update_zoomed_pk(x) %>%
+      pull(column)
+  }
+
+  pk <- pk %>%
+    map_chr(~ collapse_key_names(.x))
+
+  if (!is_empty(pk)) {
+    # FIXME: needs to change if #622 is solved
+    cat_line(trim_width(paste0("Primary key: ", pk), width))
+  }
+}
+
+collapse_key_names <- function(keys, tab = FALSE) {
+  tab <- ifelse(tab, "  ", "")
+  if (length(keys) > 1L) {
+    paste0(tab, "(", paste0(tick(keys), collapse = ", "), ")")
+  } else {
+    paste0(tab, tick(keys), collapse = ", ")
+  }
+}
 
+#' Print details about foreign keys for a given table in the `dm` object (zoomed or not)
+#' @keywords internal
+#' @noRd
+print_glimpse_table_fk <- function(x, table_name, width) {
   all_fks <- dm_get_all_fks_impl(x)
-  iwalk(table_list, function(table, table_name) {
-    cat_line("\n", trim_width(paste0("Table: ", tick(table_name)), glimpse_width))
-    pk <- dm_get_pk_impl(x, table_name) %>%
-      map_chr(~ paste0("(", paste0(tick(.x), collapse = ", "), ")"))
-    if (!is_empty(pk)) {
-      # FIXME: needs to change if #622 is solved
-      cat_line(trim_width(paste0("Primary key: ", pk), glimpse_width))
-    }
-    fk <- all_fks %>%
-      filter(child_table == table_name) %>%
-      select(-child_table) %>%
-      pmap_chr(
-        function(child_fk_cols, parent_table, parent_key_cols, on_delete) {
-          trim_width(
-            paste0(
-              "  (",
-              paste0(tick(child_fk_cols), collapse = ", "), ")",
-              " -> ",
-              "(`", paste0(parent_table, "$", parent_key_cols, collapse = "`, `"), "`) ",
-              on_delete
-            ),
-            glimpse_width
-          )
-        }
-      )
-    if (!is_empty(fk)) {
-      cat_line(length(fk), " outgoing foreign key(s):")
-      cat_line(fk)
+
+  if (is_zoomed(x)) {
+    # anticipate that some key columns could have been removed by the user
+    new_fks <- update_zoomed_fks(x, table_name, col_tracker_zoomed(x))
+
+    # if any of the key columns are missing, don't display any of the keys
+    if (nrow(new_fks) < nrow(all_fks)) {
+      return(invisible(x))
     }
-    glimpse(table, width = width, ...)
-  })
-  invisible(x)
+
+    # otherwise update the foreign keys data frame
+    all_fks <- new_fks
+  }
```

Comment:
Now we're really almost there, another minor misunderstanding: we just want to remove a composite key altogether, in case on of its columns is missing. But the FKs that will still be used once the user calls `dm_insert_zoomed()` or `dm_update_zoomed()` should actually be displayed.
So the if-condition regarding `nrow(new_fks) < nrow(all_fks)` would need to be removed.

Suggestion:
```
all_fks <- if (is_zoomed(x)) {
    # anticipate that some key columns could have been removed/renamed by the user
    update_zoomed_fks(x, table_name, col_tracker_zoomed(x))
  } else {
    dm_get_all_fks_impl(x)
  }
  ```

Reply:
Makes sense.

I think it should be in line with what you expect now:

``` r
library(dm)

dm_nycflights13() %>% dm_zoom_to(flights) %>% select(-carrier) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `flights`
#> 3 outgoing foreign key(s):
#>   `origin` -> `airports$faa` no_action
#>   `tailnum` -> `planes$tailnum` no_action
#>   (`origin`, `time_hour`) -> (`weather$origin`, `weather$time_hour`) no_action
#> Rows: 1,761
#> Columns: 18
#> $ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2…
#> $ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
#> $ day            <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ dep_time       <int> 3, 16, 450, 520, 530, 531, 535, 546, 549, 550, 553, 553…
#> $ sched_dep_time <int> 2359, 2359, 500, 525, 530, 540, 540, 600, 600, 600, 600…
#> $ dep_delay      <dbl> 4, 17, -10, -5, 0, -9, -5, -14, -11, -10, -7, -7, -7, -…
#> $ arr_time       <int> 426, 447, 634, 813, 824, 832, 1015, 645, 652, 649, 711,…
#> $ sched_arr_time <int> 437, 444, 648, 820, 829, 850, 1017, 709, 724, 703, 715,…
#> $ arr_delay      <dbl> -11, 3, -14, -7, -5, -18, -2, -24, -32, -14, -4, -33, -…
#> $ flight         <int> 727, 739, 1117, 1018, 404, 1141, 725, 380, 6055, 2114, …
#> $ tailnum        <chr> "N571JB", "N564JB", "N171US", "N35204", "N815UA", "N5EA…
#> $ origin         <chr> "JFK", "JFK", "EWR", "EWR", "LGA", "JFK", "JFK", "EWR",…
#> $ dest           <chr> "BQN", "PSE", "CLT", "IAH", "IAH", "MIA", "BQN", "BOS",…
#> $ air_time       <dbl> 183, 191, 78, 215, 210, 149, 191, 39, 48, 36, 51, 201, …
#> $ distance       <dbl> 1576, 1617, 529, 1400, 1416, 1089, 1576, 200, 229, 184,…
#> $ hour           <dbl> 23, 23, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,…
#> $ minute         <dbl> 59, 59, 0, 25, 30, 40, 40, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0…
#> $ time_hour      <dttm> 2013-01-10 23:00:00, 2013-01-10 23:00:00, 2013-01-10 0…

dm_nycflights13() %>% dm_zoom_to(flights) %>% select(-origin) %>% glimpse()
#> dm of 5 tables: `airlines`, `airports`, `flights`, `planes`, `weather`
#> 
#> Table: `flights`
#> 2 outgoing foreign key(s):
#>   `carrier` -> `airlines$carrier` no_action
#>   `tailnum` -> `planes$tailnum` no_action
#> Rows: 1,761
#> Columns: 18
#> $ year           <int> 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2013, 2…
#> $ month          <int> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1…
#> $ day            <int> 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,…
#> $ dep_time       <int> 3, 16, 450, 520, 530, 531, 535, 546, 549, 550, 553, 553…
#> $ sched_dep_time <int> 2359, 2359, 500, 525, 530, 540, 540, 600, 600, 600, 600…
#> $ dep_delay      <dbl> 4, 17, -10, -5, 0, -9, -5, -14, -11, -10, -7, -7, -7, -…
#> $ arr_time       <int> 426, 447, 634, 813, 824, 832, 1015, 645, 652, 649, 711,…
#> $ sched_arr_time <int> 437, 444, 648, 820, 829, 850, 1017, 709, 724, 703, 715,…
#> $ arr_delay      <dbl> -11, 3, -14, -7, -5, -18, -2, -24, -32, -14, -4, -33, -…
#> $ carrier        <chr> "B6", "B6", "US", "UA", "UA", "AA", "B6", "B6", "EV", "…
#> $ flight         <int> 727, 739, 1117, 1018, 404, 1141, 725, 380, 6055, 2114, …
#> $ tailnum        <chr> "N571JB", "N564JB", "N171US", "N35204", "N815UA", "N5EA…
#> $ dest           <chr> "BQN", "PSE", "CLT", "IAH", "IAH", "MIA", "BQN", "BOS",…
#> $ air_time       <dbl> 183, 191, 78, 215, 210, 149, 191, 39, 48, 36, 51, 201, …
#> $ distance       <dbl> 1576, 1617, 529, 1400, 1416, 1089, 1576, 200, 229, 184,…
#> $ hour           <dbl> 23, 23, 5, 5, 5, 5, 5, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6, 6,…
#> $ minute         <dbl> 59, 59, 0, 25, 30, 40, 40, 0, 0, 0, 0, 0, 0, 0, 0, 5, 0…
#> $ time_hour      <dttm> 2013-01-10 23:00:00, 2013-01-10 23:00:00, 2013-01-10 0…
```

<sup>Created on 2022-06-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
Looks great to me now, thank you!